### DISCUSSION TITLE
Plugin with id 'org.teavm' not found? (Gradle)

### DISCUSSION THREAD

**Discussion by WorldEditAxe** [konsoletyper/teavm#738] [Q&A] [answered]
Hi, I am attempting to add TeaVM into my Gradle project using the following Gradle configuration:
```gradle
buildscript {
    repositories {
        maven {
            url = uri("https://plugins.gradle.org/m2/")
        }
    }
    dependencies {
        classpath("org.teavm:teavm-gradle-plugin:0.8.1") {
            attributes {
                attribute(Attribute.of("org.gradle.docstype", String), "sources")
            }
        }
    }
}

apply plugin: 'org.teavm'
```
However, Gradle fails to recognize the plugin, and errors out with `Plugin with id 'org.teavm' not found.` Is there anything wrong with my configuration? I have directly copied the installation instructions from the [Gradle plugin repository](https://plugins.gradle.org/plugin/org.teavm) apart from a few modifications to resolve a Gradle variation ambiguity issue.
Also, does TeaVM support threads for WASM targets? If so, does TeaVM support real or green threads for WASM targets?

**Comment by konsoletyper** [comment]
Did you try [these instructions](https://teavm.org/docs/tooling/gradle.html)? Does legacy plugin application works for other plugins? Which Gradle version are you using? Which Java version? Could you provide Gradle output with `--stacktrace` option specified?

**Reply by WorldEditAxe** [reply]
1. No – refer to second question for reason.
2. Yes. I am using the legacy method of plugin importing due to other plugins being imported the same way.
   * Using the method as listed on the website results in the following error I'm not sure how to fix (not very fluent in Gradle):
```
> Could not resolve all artifacts for configuration ':classpath'.
   > Could not resolve org.teavm:teavm-gradle-plugin:0.8.1.
     Required by:
         project : > org.teavm:org.teavm.gradle.plugin:0.8.1
      > Cannot choose between the following variants of org.teavm:teavm-gradle-plugin:0.8.1:
          - javadocElements
          - sourcesElements
        All of them match the consumer attributes:
          - Variant 'javadocElements' capability org.teavm:teavm-gradle-plugin:0.8.1:
              - Unmatched attributes:
                  - Found org.gradle.category 'documentation' but wasn't required.
                  - Found org.gradle.docstype 'javadoc' but wasn't required.
                  - Required org.gradle.jvm.version '8' but no value provided.
                  - Required org.gradle.libraryelements 'jar' but no value provided.
                  - Found org.gradle.status 'release' but wasn't required.
              - Compatible attributes:
                  - Required org.gradle.dependency.bundling 'external' and found compatible value 'external'.
                  - Required org.gradle.usage 'java-runtime' and found compatible value 'java-runtime'.
          - Variant 'sourcesElements' capability org.teavm:teavm-gradle-plugin:0.8.1:
              - Unmatched attributes:
                  - Found org.gradle.category 'documentation' but wasn't required.
                  - Found org.gradle.docstype 'sources' but wasn't required.
                  - Required org.gradle.jvm.version '8' but no value provided.
                  - Required org.gradle.libraryelements 'jar' but no value provided.
                  - Found org.gradle.status 'release' but wasn't required.
              - Compatible attributes:
                  - Required org.gradle.dependency.bundling 'external' and found compatible value 'external'.
                  - Required org.gradle.usage 'java-runtime' and found compatible value 'java-runtime'.
```
4. Gradle 6.0
5. IntelliJ-installed OpenJDK Amazon Corretto 1.8.0_382
6. https://pastebin.com/1yqAXWS2

**Comment by konsoletyper** [comment]
WASM targets are experimental and not recommended to use. Consider JS target instead. Threads in all the targets are emulated via green threads.

**Reply by WorldEditAxe** [reply]
1. I understand, but I'd still like to utilize WASM for performance reasons. Is there anything I should be aware of/know about WASM support?
2. Are there plans to use real threads for the WASM target/platform? I believe Emscripten supports POSIX threads [via SharedArrayBuffers.](https://emscripten.org/docs/porting/pthreads.html)
3. What's this [C target](https://github.com/konsoletyper/teavm/blob/83ace9bf3e697ab65393f0ddb9370de5b38382a4/tools/core/src/main/java/org/teavm/tooling/TeaVMTargetType.java#L4) about?

**Comment by konsoletyper** [comment, marked as answer]
> IntelliJ-installed OpenJDK Amazon Corretto 1.8.0_382

TeaVM requires Java 11+

> I am using the legacy method of plugin importing due to other plugins being imported the same way. 

You better try new plugin DSL.

> Gradle 6.0

Never tested TeaVM with this ancient Gradle version. I'd recommend to use at least 7.0

**Reply by WorldEditAxe** [reply]
I've updated Java (to 11), used the new DSL/suggested method, and updated Gradle to 7.0. Instead of the previous error, I'm getting a new one: https://pastebin.com/67XhYTyS
I apologize in advance, as I am not very experienced in Gradle.

**Reply by konsoletyper** [reply]
You should apply language plugin (e.g. Java or Kotlin) *before* TeaVM plugin.

**Reply by WorldEditAxe** [reply]
I was able to get everything up and running, thanks! One last question: How can I get the compiled JavaScript files to end up in outputDir instead of `${outputDir}/js`?
Also, has the JavaScript file structure changed? `eval()`ing the JavaScript file throws the following error:
```
Uncaught TypeError: Cannot read properties of undefined (reading 'prototype')
    at eval (eval at xhr.onreadystatechange ((index):118:9), <anonymous>:4:301)
    at eval (eval at xhr.onreadystatechange ((index):118:9), <anonymous>:63117:3)
    at xhr.onreadystatechange ((index):135:9)
```
This has not happened in the past with TeaVM versions 0.6.1 and 0.7.0. More specifically, the error occurred when a value was assigned to `$rt_globals.Array.prototype.fill`.

**Reply by konsoletyper** [reply]
> How can I get the compiled JavaScript files to end up in outputDir instead of `${outputDir}/js`?

Sorry, forgot to document this. You should use `relativePathInOutputDir` - it's `js` by default in JS target configuration. Just specify empty string instead.

> Also, has the JavaScript file structure changed?

It's constantly changing according to specific needs.

> eval()ing the JavaScript file throws the following error:

Never tried to `eval` TeaVM output. For what purpose do you do that?

> This has not happened in the past with TeaVM versions 0.6.1 and 0.7.0.

BTW, if you want to try WebAssembly, you can try latest [preview builds](https://teavm.org/docs/intro/preview-builds.html).

**Reply by WorldEditAxe** [reply]
Regarding the preview builds: Thank you, I will check out the builds later.
In response to your question: My site's landing page fetches the .js file from the server and stores it in memory. It is then used to launch the code on both the main/rendering thread and any worker threads. `blob:` URLs are passed to those workers to launch those them while minimizing requests sent to the server. To launch the code on the render thread, I previously used `eval()`.

**Reply by konsoletyper** [reply]
Alternative approach is to use `new Function(...)`. Then you can call `apply` on this function with proper this (global) specified. I guess this may be related to scoping in eval.

**Reply by konsoletyper** [reply]
> are passed to those workers to launch those them while minimizing requests sent to the server

I don't think it's the case. Browser may store script in the cache.

**Reply by WorldEditAxe** [reply]
That more or less depends on the caching policy, though. I have disabled that on my website entirely due to the issues it caused after rolling out site updates.
Either way, thank you!

**Reply by konsoletyper** [reply]
Sure. That's why you should learn how to set up caching policies instead of reinventing the wheel. I'd recommend following approach: each time you update static script, put it into separate directory (usually version or build number) and update HTML page (which is not cached, or cached only very limited time) to get script from this new directory. That's how people usually do.

**Comment by konsoletyper** [comment]
> I understand, but I'd still like to utilize WASM for performance reasons. Is there anything I should be aware of/know about WASM support?

The problem with "experimental" status is that I don't have any documentation for Wasm yet, and have no time to write it.

> Are there plans to use real threads for the WASM target/platform? I believe Emscripten supports POSIX threads [via SharedArrayBuffers.](https://emscripten.org/docs/porting/pthreads.html)

No, there are no such plans. Multithreaded code is extremely hard to develop and test, I simple have not enough resources for it. Another reason: TeaVM historically written for JS, which does not support multiple threads, so the whole Java library was implemented with green threads in mind. To introduce real threads, I have to review the entire library and find parts that need to support threads. Also, these parts can't be simply rewritten, since JS target should keep working, so I need to invent a mechanism that allows to have platform threads and green threads implementations.

> What's this [C target](https://github.com/konsoletyper/teavm/blob/83ace9bf3e697ab65393f0ddb9370de5b38382a4/tools/core/src/main/java/org/teavm/tooling/TeaVMTargetType.java#L4) about?

It's about generation of C code that can be compiled into native code. For example, [jbox2 sample](https://github.com/konsoletyper/teavm/tree/83ace9bf3e697ab65393f0ddb9370de5b38382a4/samples/benchmark) has GTK+ implementation that can be compiled into native binary in Linux.
